UI improvements and chat channel functionality

joeShuff opened the pull request:
a few changes made in this branch:

### Add game config to the pre-game gameboard.
![image](https://user-images.githubusercontent.com/15524362/119881790-9a65f380-bf25-11eb-86d5-abe833cf0f25.png)

### Add configurable progress notifications. Support for progress on lines or number of items (Full card only)
example config
```
notify-progress:
  lines:
    - each
  full-count:
    - 10
    - 20
    - 24
```
![image](https://user-images.githubusercontent.com/15524362/119881711-83bf9c80-bf25-11eb-839e-770972460586.png)

### Stop gameboard flickering, most noticable with timer

|       BEFORE       |        AFTER      |
| :----------------: | :----------------: |
| ![timer-old](https://user-images.githubusercontent.com/15524362/119883110-10b72580-bf27-11eb-9387-ee056978cb6d.gif) | ![timer-new](https://user-images.githubusercontent.com/15524362/119881936-c2555700-bf25-11eb-92ab-91743b506963.gif) |

### Join team quicker
Added support to add yourself to a team with `/team <teamcolor>`. 

I added this because playing a private game with friends is easier to have everyone adding themselves to teams and we found it quite annoying typing out the whole add command. Especially without auto completion.


### Still to do
I don't full understand how the lockout gamemode works as when you enable it the `numOfLinesToComplete` is set to 0 which shows the wincondition as 0 lines. _How would you expect lockout to display on the pregame game board?_

_Could the progress notifications work with the lockout gamemode?_

Extremelyd1 commented:
To answers the questions you still have, let me preface this comment by explaining the lockout gamemode: after a certain (configurable) number of teams collect an item it will be locked and can't be collected by the remaining teams. Thus the win condition for lockout is dynamically calculated: if a team has obtained a number of items `x` and all opposing teams cannot reach more than `x` items, that team has won.

> How would you expect lockout to display on the pregame game board?

The entries "Win Condition" and "Lockout" on the pregame game board can be condensed in "Game Mode" or "Game Type" which has a potential values: "[x] Lines", "Full Card", or "Lockout".

> Could the progress notifications work with the lockout gamemode?

Definitely! While there are no lines to complete in lockout, you can still post progress notification for the number of collected items.

joeShuff commented:
Ok that makes sense now, thanks for explaining! I will make some changes now I understand it.

joeShuff commented:
I remembered another quality of life change I wanted to try to implement so I'll get that added before review

joeShuff commented:
more new features:

### Chat Channels
When you're doing solo teams, team chatting being the default can be annoying so I added the `\channel` command to allow people to switch between TEAM and GLOBAL default chatting.

Because of this I needed to also add a command for talking to your team if you're default in global, so I also added a `\tchat` command and a `\tloc` command to quickly send your location to your team or note down coords that are important to you for later.

![ezgif com-gif-maker(2)](https://user-images.githubusercontent.com/15524362/120031789-2abd3a80-bff1-11eb-911b-a5e237c72584.gif)

Extremelyd1 commented:
Thanks for your interest in this project and contributing to it, but as a heads up for next time: please try to limit a pull request to a single feature. While this pull request is still manageable in terms of reviewing, it would greatly speed up review and interactions if it only pertained to a single feature.

Now content-wise, I already have a few remarks:
First of all, there already exists a command to share coordinates to you team, namely `/coordinates`, aliased as `/coord` or `/coords`.
Secondly, Perhaps this is a bit nitpicky, but I would like some commands to change names:
I don't find the name for the `/tchat` command intuitive, I would prefer something like `/teamchat` with alias `/t`. This alias, while it is short and convenient to type chat messages to your team with, might resemble the `/team` command too much. So perhaps the alias could also be `/tc`.
Adding to this, the `/team` command used for joining teams directly might be better renamed (or aliased) to `/join`.

joeShuff commented:
Sorry about that, I got too carried away fixing little things that bothered me and thought they were too small to warrant their own PR. I'll keep that in mind for if I create another.

> First of all, there already exists a command to share coordinates

I completely missed that, I'll remove my implementation.

> Secondly, Perhaps this is a bit nitpicky, but I would like some commands to change names

I do agree with your comments and have changed teamchat to be `/teamchat` with alias `/tc` and created a new `/join` command to reduce overcomplicating the existing team command.

Extremelyd1 commented on src/main/java/com/extremelyd1/command/ChannelCommand.java:
```diff
@@ -0,0 +1,78 @@
+            return true;
+        }
+
+        ChatChannelController.ChatChannel channel = ChatChannelController.ChatChannel.valueOf(args[0].toUpperCase());
```
`valueOf` on enum types will throw an `IllegalArgumentException` if the given enum constant can not be found. The if-statement below checking for `channel != null` should thus be replaced by a try-catch instead.

joeShuff replied:
You are right, I am used to Kotlin where there is a safe option for this. I have modified the method to use a try catch.

Extremelyd1 commented on src/main/java/com/extremelyd1/command/ChannelCommand.java:
```diff
@@ -0,0 +1,78 @@
+    @Override
+    public List<String> onTabComplete(CommandSender sender, Command command, String s, String[] args) {
+        if (!(sender instanceof Player)) {
+            if (!sender.isOp()) return emptyList();
```
I'm not sure what this check is for. If the `CommandSender` is not a `Player`, is it still necessary to check whether they are  an operator? I might be mistaken, but tab completion is only relevant for `Player` instances isn't it?

joeShuff replied:
you're correct, I think this was just some copy paste error. Resolved.

Extremelyd1 commented on src/main/java/com/extremelyd1/command/JoinCommand.java:
```diff
@@ -0,0 +1,85 @@
+        }
+
+        PlayerTeam checkTeam = game.getTeamManager().getTeamByName(args[0]);
+        if (checkTeam != null && sender instanceof Player) {
```
Why does this if-statement include a clause for whether the `CommandSender` is a `Player`? I don't think this command should be executable for anything other than players. Thus the first line of this method can simply catch that by changing the boolean parameter to `true`. Moreover, the outputted message in the else-statement is non-sensical if the team is correct, but the sender is a non-player class.

joeShuff replied:
I felt unsafe casting the sender to a player without checking it first. I have amended the first line to not allow console sender and removed the check.

Extremelyd1 commented on src/main/java/com/extremelyd1/command/JoinCommand.java:
```diff
@@ -0,0 +1,85 @@
+
+        if (args.length > 1) return teams;
+
+        for (PlayerTeam team: game.getTeamManager().getAvailableTeams()) {
```
There is a space missing between `team` and the colon `:`.

joeShuff replied:
added space

Extremelyd1 commented on src/main/java/com/extremelyd1/command/JoinCommand.java:
```diff
@@ -0,0 +1,85 @@
+import org.bukkit.entity.Player;
+
+import java.util.ArrayList;
+import java.util.Arrays;
```
This import is redundant.

joeShuff replied:
Optimized imports.

Extremelyd1 commented on src/main/java/com/extremelyd1/command/ChannelCommand.java:
```diff
@@ -0,0 +1,78 @@
+import org.bukkit.command.TabExecutor;
+import org.bukkit.entity.Player;
+
+import java.util.ArrayList;
```
This import is redundant. This happens a few more times, so perhaps you can simply optimize imports in IntelliJ or a similar command in your preferred IDE for the files that were changed.

joeShuff replied:
Ran optimize imports on all modified files.

Extremelyd1 commented on src/main/java/com/extremelyd1/command/JoinCommand.java:
```diff
@@ -0,0 +1,85 @@
+        if (args.length > 1) return teams;
+
+        for (PlayerTeam team: game.getTeamManager().getAvailableTeams()) {
+            if (team.getName().toLowerCase().contains(args[0].toLowerCase())) {
```
I'm not too familiar with the `TabExecutor` interface and this overridden method, but doesn't this throw a `NullPointerException` if the `args` parameter is an empty array?

joeShuff replied:
the `args` param always has at least 1 entry in this case but it can be empty when the user hasn't typed anything. However I agree it's not safe so I've modified the above length check to be `!=` instead of `>`

Extremelyd1 commented on src/main/java/com/extremelyd1/command/ChannelCommand.java:
```diff
@@ -0,0 +1,78 @@
+
+    @Override
+    public boolean onCommand(CommandSender sender, Command command, String s, String[] args) {
+        if (!(sender instanceof Player)) {
```
Please use `CommandUtil#checkCommandSender` for this to unify the checks and the corresponding error messages.

joeShuff replied:
Added this

Extremelyd1 commented on src/main/java/com/extremelyd1/command/TeamChatCommand.java:
```diff
@@ -0,0 +1,46 @@
+
+    @Override
+    public boolean onCommand(CommandSender sender, Command command, String s, String[] strings) {
+        if (!(sender instanceof Player)) {
```
Please use `CommandUtil#checkCommandSender` for this to unify the checks and the corresponding error messages.

joeShuff replied:
changed

Extremelyd1 commented on src/main/java/com/extremelyd1/command/TeamChatCommand.java:
```diff
@@ -0,0 +1,46 @@
+package com.extremelyd1.command;
+
+import com.extremelyd1.game.Game;
+import com.extremelyd1.game.team.PlayerTeam;
```
This import is redundant.

joeShuff replied:
optimized imports

Extremelyd1 commented on src/main/java/com/extremelyd1/game/Game.java:
```diff
@@ -125,6 +126,11 @@
+    /**
+     * Chat Channel controller
+     */
+    private ChatChannelController chatChannelController;
```
This can be `final`.

joeShuff replied:
made `final`

Extremelyd1 commented on src/main/java/com/extremelyd1/game/Game.java:
```diff
@@ -507,6 +518,8 @@ public void onMaterialCollected(Player player, Material material) {
 
         PlayerTeam collectorTeam = (PlayerTeam) team;
 
+        int linesPreCheck = bingoCard.getNumLinesComplete(collectorTeam);
```
Perhaps this variable is better named "linesCompletedBefore", which feels more descriptive.

joeShuff replied:
changed to what you recommended, variable names are always the hardest part of coding!

Extremelyd1 commented on src/main/java/com/extremelyd1/game/Game.java:
```diff
@@ -530,6 +543,39 @@ public void onMaterialCollected(Player player, Material material) {
                 // for other teams
             }
 
+            if (winConditionChecker.isFullCard()) {
```
This entire logic might be moved to the `ProgressController`, so this method doesn't get too bloated. As far as I'm aware, it doesn't need a lot of information and simply conditionally outputs/broadcasts a message.

joeShuff replied:
I've moved the logic into the `ProgressController`

Extremelyd1 commented on src/main/java/com/extremelyd1/game/Game.java:
```diff
@@ -530,6 +543,39 @@ public void onMaterialCollected(Player player, Material material) {
             }
 
+            if (winConditionChecker.isFullCard()) {
+                if (config.getProgressController().shouldNotifyCountAmount(collectorTeam.getNumCollected())) {
```
The logic for full card and lockout game modes are exactly the same. The comment below suggests that this perhaps shouldn't be, in which case this comment can be disregarded.

joeShuff replied:
the next comment is right it was incorrect.

Extremelyd1 commented on src/main/java/com/extremelyd1/game/chat/ChatChannelController.java:
```diff
@@ -0,0 +1,30 @@
+public class ChatChannelController {
+
+    public enum ChatChannel {
+        TEAM(), GLOBAL()
```
The parenthesis are unnecessary and while there is no style guide, I would appreciate it if each enum constant is on its own line and that the enum definition is declared at the bottom of the class.

joeShuff replied:
Made the changes you requested

Extremelyd1 commented on src/main/java/com/extremelyd1/game/chat/ChatChannelController.java:
```diff
@@ -0,0 +1,30 @@
+    public Map<String, ChatChannel> playerChatChannels = new HashMap<>();
+
+    public void setPlayerChatChannel(Player player, ChatChannel channel) {
+        playerChatChannels.put(player.getUniqueId().toString(), channel);
```
Is using the `String` representation of the UUID better than using the UUID object itself as a key in a HashMap? I believe UUID overrides the `hashCode` and `equals` methods, so it would be perfectly fine to use as a key.

joeShuff replied:
ah yeah it does override those methods, I wasn't sure if it was a safe key. Changed the key to the UUID.

Extremelyd1 commented on src/main/java/com/extremelyd1/game/chat/ChatChannelController.java:
```diff
@@ -0,0 +1,30 @@
+
+    public ChatChannel getPlayerChatChannel(Player player) {
+        ChatChannel channel = playerChatChannels.get(player.getUniqueId().toString());
+        if (channel == null) {
```
Another remark about the style: you can use the `Map#containsKey` method instead of checking for `null`. In some cases          `null` values are allowed in a map and then this would be an issue, because you can't differentiate between whether there exists no mapping or it explicitly maps to `null`.

joeShuff replied:
of course, again this is my familiarity with null safety in Kotlin making me forget Java ways! I have modified this method.

Extremelyd1 commented on src/main/java/com/extremelyd1/game/progress/ProgressController.java:
```diff
@@ -0,0 +1,45 @@
+    /**
+     * Rules for lines notifications
+     */
+    private List<String> linesRules;
```
This (and all other member variables in this class) can be `final`.

joeShuff replied:
done

Extremelyd1 commented on src/main/java/com/extremelyd1/game/progress/ProgressController.java:
```diff
@@ -0,0 +1,45 @@
+        return fullCardCountRules.contains(String.valueOf(itemsCollected));
+    }
+
+    public Boolean shouldNotifyLockoutAmount(int itemsCollected) {
```
This method is unused, which probably shouldn't be the case.

joeShuff replied:
This was a mistake elsewhere which has been resolved

Extremelyd1 commented on src/main/java/com/extremelyd1/game/progress/ProgressController.java:
```diff
@@ -0,0 +1,45 @@
+        lockoutRules = config.getStringList("notify-progress.lockout");
+    }
+
+    private Boolean notifyEachLine() {
```
Preferably use the primitive `boolean` instead of the object wrapper, unless absolutely necessary. This holds for the other methods in this class as well.

joeShuff replied:
done

Extremelyd1 commented on src/main/java/com/extremelyd1/gameboard/GameBoard.java:
```diff
@@ -61,7 +61,19 @@ public GameBoard(Game game) {
         // Clear all existing scores
         for (String entry : this.scoreboard.getEntries()) {
-            this.scoreboard.resetScores(entry);
+            Boolean entryStillExists = false;
```
Again, please use the primitive `boolean` instead of the object wrapper.

joeShuff replied:
done

Extremelyd1 commented on src/main/java/com/extremelyd1/gameboard/GameBoard.java:
```diff
@@ -61,7 +61,19 @@ public GameBoard(Game game) {
-            this.scoreboard.resetScores(entry);
+            Boolean entryStillExists = false;
+
+            for (int i = 0; i < boardEntries.size(); i++) {
```
There is no reason to use an indexed for loop here, `boardEntries` is a `List` and implements `Iterable`.

joeShuff replied:
changed

Extremelyd1 commented on src/main/java/com/extremelyd1/gameboard/GameBoard.java:
```diff
@@ -61,7 +61,19 @@ public GameBoard(Game game) {
+                BoardEntry newEntry = boardEntries.get(i);
+
+                if (newEntry.getString().equals(entry)) {
+                    entryStillExists = true;
```
After this statement you can break out of the for-loop.

joeShuff replied:
done

Extremelyd1 commented on src/main/java/com/extremelyd1/gameboard/GameBoard.java:
```diff
@@ -61,7 +61,19 @@ public GameBoard(Game game) {
     public void update() {
         // Clear all existing scores
```
This comment is not relevant anymore, perhaps document the update logic a bit more. While I can understand it, it might not be as clear for people unfamiliar with the codebase.

joeShuff replied:
Updated the java docs to explain the process and updated this comment

Extremelyd1 commented on src/main/java/com/extremelyd1/listener/ChatListener.java:
```diff
@@ -32,10 +33,16 @@ public void onAsyncPlayerChat(AsyncPlayerChatEvent e) {
                     player.getName() + ": " + e.getMessage()
             );
+        }
+        else if (chatChannel == ChatChannelController.ChatChannel.GLOBAL) {
```
The else-if-statement can be placed after the closing curly brace of the previous line.

joeShuff replied:
done

Extremelyd1 commented on src/main/java/com/extremelyd1/game/Game.java:
```diff
@@ -530,6 +543,39 @@ public void onMaterialCollected(Player player, Material material) {
+                    }
+                }
+            } else if (winConditionChecker.getCompletionsToLock() > 0) {
+                if (config.getProgressController().shouldNotifyCountAmount(collectorTeam.getNumCollected())) {
```
Shouldn't this use a different method in `ProgressController`?
```suggestion
                if (config.getProgressController().shouldNotifyLockoutAmount(collectorTeam.getNumCollected())) {
```

joeShuff replied:
this has been moved into the controller and change has been made.

Extremelyd1 reviewed (changes requested):
I have simply looked through the code and commented on things that should be changed, I haven't play-tested the changes yet.
Some inconsistencies and some stylistic choices that I don't agree with, but all-in-all well done!

joeShuff reviewed (commented):
Made the changes you requested 😄

joeShuff commented:
I'm not sure I responded to your requests correctly :/ It looks like I reviewed my own code, still getting used to githubs review system. I hope you can see my responses easily and my changes.

Extremelyd1 commented:
I believe you created a review containing the comments that were responses on the threads I created. To avoid this in the future, you can just respond to the threads I create with comments (by selecting "Add single comment" instead of "Start review"). This will add the comment immediately on the thread, so perhaps you want to add these comments after you have pushed the necessary commits.
